modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/exec/Outbox.java

```diff
@@ -0,0 +1,277 @@
+            perNodeBuffers.computeIfAbsent(node, this::createBuffer).end();
+
+        close();
+    }
+
+    /** {@inheritDoc} */
+    @Override public void close() {
+        context().mailboxRegistry().unregister(this);
+    }
+
+    /** */
+    private void sendBatch(UUID nodeId, int batchId, List<?> rows) {
+        context().exchange().sendBatch(this, nodeId, queryId(), targetFragmentId, exchangeId, batchId, rows);
+    }
+
+    /** */
+    private void sendCancel(UUID nodeId, int batchId) {
+        context().exchange().cancel(this, nodeId, queryId(), targetFragmentId, exchangeId, batchId);
+    }
+
+    /** */
+    private Buffer createBuffer(UUID nodeId) {
+        return new Buffer(nodeId, this);
+    }
+
+    /** */
+    private static final class Buffer {
+        /** */
+        private final Outbox<?> owner;
+
+        /** */
+        private final UUID nodeId;
+
+        /** */
+        private int hwm = -1;
+
+        /** */
+        private int lwm = -1;
+
+        /** */
+        private List<Object> curr;
+
+        /** */
+        private Buffer(UUID nodeId, Outbox<?> owner) {
+            this.nodeId = nodeId;
+            this.owner = owner;
+
+            curr = new ArrayList<>(ExchangeService.BATCH_SIZE + 1); // extra space for end marker;
+        }
+
+        /**
+         * Adds a row to current batch.
+         *
+         * @param row Row.
+         */
+        public void add(Object row) {
+            assert ready();
+
+            if (curr.size() == ExchangeService.BATCH_SIZE) {
+                owner.sendBatch(nodeId, ++hwm, curr);
+                curr = new ArrayList<>(ExchangeService.BATCH_SIZE + 1); // extra space for end marker;
+            }
+
+            curr.add(row);
+        }
+
+        /**
+         * Signals data is over.
+         */
+        public void end() {
+            int batchId = hwm + 1;
+            hwm = Integer.MAX_VALUE;
+
+            List<Object> tmp = curr;
+            curr = null;
+
+            tmp.add(EndMarker.INSTANCE);
+            owner.sendBatch(nodeId, batchId, tmp);
+        }
+
+        public void cancel() {
+            int batchId = hwm + 1;
+            hwm = Integer.MAX_VALUE;
+
+            curr = null;
+            owner.sendCancel(nodeId, batchId);
+        }
+
+        /**
+         * Checks whether there is a place for a new row.
+         *
+         * @return {@code True} is it possible to add a row to a batch.
+         */
+        private boolean ready() {
+            assert curr != null;
+
+            return hwm != Integer.MAX_VALUE
+                && hwm - lwm < ExchangeService.PER_NODE_BATCH_COUNT
+                || curr.size() < ExchangeService.BATCH_SIZE;
+        }
+
+        /**
+         * Callback method.
+         *
+         * @param id batch ID.
+         */
+        private void onAcknowledge(int id) {
+            if (lwm > id)
+                return;
+
+            boolean request = hwm - lwm == ExchangeService.PER_NODE_BATCH_COUNT;
```

Comment:
As a contract says, in case a node returns false as an answer of push call it has to request data as soon as it's able to consume rows. The fact the condition is true means the node returned false, so it should request data.

Reply:
It's there not to generate unnecessary request calls